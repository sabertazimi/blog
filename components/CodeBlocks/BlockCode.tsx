import type { Language } from 'prism-react-renderer'
import cx from 'classnames'
import Highlight, { defaultProps } from 'prism-react-renderer'
import styles from './BlockCode.module.css'
import theme from './monokai'

interface Props {
  enableLine: boolean
  lines: Set<number>
  language: Language
  children: string
}

function BlockCode({
  enableLine,
  lines,
  language,
  children,
}: Props): JSX.Element {
  return (
    <Highlight
      {...defaultProps}
      code={children}
      language={language}
      theme={theme}
    >
      {({ className, style, tokens, getLineProps, getTokenProps }) => (
        <pre className={cx(className, styles.code)} style={style}>
          {tokens.map((line, index) => (
            <div
              {...getLineProps({ line, key: index })}
              className={cx(
                'token-line',
                lines.has(index + 1) ? styles.highlight : styles.hover,
              )}
              // eslint-disable-next-line react/no-array-index-key -- No other unique key.
              key={index}
            >
              {enableLine
                ? (
                    <span className={styles.number} key={0}>{index + 1}</span>
                  )
                : (
                    <span className={styles.placeholder} key={0}></span>
                  )}
              {line.map((token, key) => (
                // eslint-disable-next-line react/no-array-index-key -- No other unique key.
                <span {...getTokenProps({ token, key })} key={key + Number(enableLine)} />
              ))}
            </div>
          ))}
        </pre>
      )}
    </Highlight>
  )
}

export default BlockCode
