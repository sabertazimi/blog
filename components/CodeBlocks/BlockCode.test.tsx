import { render } from '@testing-library/react';
import BlockCode from './BlockCode';

describe('BlockCode', () => {
  const languages = [
    'html',
    'xml',
    'yml',
    'yaml',
    'css',
    'json',
    'md',
    'markdown',
    'js',
    'javascript',
    'ts',
    'typescript',
    'coffee',
    'coffeescript',
    'jsx',
    'tsx',
    'objc',
    'objectivec',
    'vue',
    'go',
    'java',
    'rust',
  ];

  test.each(languages)(
    'should render different language correctly (snapshot)',
    language => {
      const { container } = render(
        <BlockCode className={`language-${language}`}>
          const foo = bar();
        </BlockCode>
      );

      expect(container).toMatchSnapshot();
    }
  );
});
