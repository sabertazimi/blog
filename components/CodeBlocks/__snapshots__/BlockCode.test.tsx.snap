// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BlockCode should render different language correctly (snapshot) 1`] = `
<div>
  <pre
    class="prism-code language-html code"
    style="color: rgb(118, 217, 230); background-color: rgb(42, 42, 42);"
  >
    <div
      class="token-line hover"
      style="color: rgb(118, 217, 230);"
    >
      <span
        class="number"
      >
        1
      </span>
      <span
        class="token plain"
      >
        const foo = bar();
      </span>
    </div>
  </pre>
</div>
`;

exports[`BlockCode should render different language correctly (snapshot) 2`] = `
<div>
  <pre
    class="prism-code language-xml code"
    style="color: rgb(118, 217, 230); background-color: rgb(42, 42, 42);"
  >
    <div
      class="token-line hover"
      style="color: rgb(118, 217, 230);"
    >
      <span
        class="number"
      >
        1
      </span>
      <span
        class="token plain"
      >
        const foo = bar();
      </span>
    </div>
  </pre>
</div>
`;

exports[`BlockCode should render different language correctly (snapshot) 3`] = `
<div>
  <pre
    class="prism-code language-yml code"
    style="color: rgb(118, 217, 230); background-color: rgb(42, 42, 42);"
  >
    <div
      class="token-line hover"
      style="color: rgb(118, 217, 230);"
    >
      <span
        class="number"
      >
        1
      </span>
      <span
        class="token plain"
      >
        const foo = bar();
      </span>
    </div>
  </pre>
</div>
`;

exports[`BlockCode should render different language correctly (snapshot) 4`] = `
<div>
  <pre
    class="prism-code language-yaml code"
    style="color: rgb(118, 217, 230); background-color: rgb(42, 42, 42);"
  >
    <div
      class="token-line hover"
      style="color: rgb(118, 217, 230);"
    >
      <span
        class="number"
      >
        1
      </span>
      <span
        class="token plain"
      >
        const foo = bar();
      </span>
    </div>
  </pre>
</div>
`;

exports[`BlockCode should render different language correctly (snapshot) 5`] = `
<div>
  <pre
    class="prism-code language-css code"
    style="color: rgb(118, 217, 230); background-color: rgb(42, 42, 42);"
  >
    <div
      class="token-line hover"
      style="color: rgb(118, 217, 230);"
    >
      <span
        class="number"
      >
        1
      </span>
      <span
        class="token plain"
      >
        const foo = 
      </span>
      <span
        class="token function"
        style="color: rgb(166, 226, 45);"
      >
        bar
      </span>
      <span
        class="token punctuation"
        style="color: rgb(248, 248, 242);"
      >
        (
      </span>
      <span
        class="token punctuation"
        style="color: rgb(248, 248, 242);"
      >
        )
      </span>
      <span
        class="token punctuation"
        style="color: rgb(248, 248, 242);"
      >
        ;
      </span>
    </div>
  </pre>
</div>
`;

exports[`BlockCode should render different language correctly (snapshot) 6`] = `
<div>
  <pre
    class="prism-code language-json code"
    style="color: rgb(118, 217, 230); background-color: rgb(42, 42, 42);"
  >
    <div
      class="token-line hover"
      style="color: rgb(118, 217, 230);"
    >
      <span
        class="number"
      >
        1
      </span>
      <span
        class="token plain"
      >
        const foo = bar();
      </span>
    </div>
  </pre>
</div>
`;

exports[`BlockCode should render different language correctly (snapshot) 7`] = `
<div>
  <pre
    class="prism-code language-md code"
    style="color: rgb(118, 217, 230); background-color: rgb(42, 42, 42);"
  >
    <div
      class="token-line hover"
      style="color: rgb(118, 217, 230);"
    >
      <span
        class="number"
      >
        1
      </span>
      <span
        class="token plain"
      >
        const foo = bar();
      </span>
    </div>
  </pre>
</div>
`;

exports[`BlockCode should render different language correctly (snapshot) 8`] = `
<div>
  <pre
    class="prism-code language-markdown code"
    style="color: rgb(118, 217, 230); background-color: rgb(42, 42, 42);"
  >
    <div
      class="token-line hover"
      style="color: rgb(118, 217, 230);"
    >
      <span
        class="number"
      >
        1
      </span>
      <span
        class="token plain"
      >
        const foo = bar();
      </span>
    </div>
  </pre>
</div>
`;

exports[`BlockCode should render different language correctly (snapshot) 9`] = `
<div>
  <pre
    class="prism-code language-js code"
    style="color: rgb(118, 217, 230); background-color: rgb(42, 42, 42);"
  >
    <div
      class="token-line hover"
      style="color: rgb(118, 217, 230);"
    >
      <span
        class="number"
      >
        1
      </span>
      <span
        class="token keyword"
        style="color: rgb(249, 38, 114);"
      >
        const
      </span>
      <span
        class="token plain"
      >
         foo 
      </span>
      <span
        class="token operator"
        style="color: rgb(167, 122, 254);"
      >
        =
      </span>
      <span
        class="token plain"
      >
         
      </span>
      <span
        class="token function"
        style="color: rgb(166, 226, 45);"
      >
        bar
      </span>
      <span
        class="token punctuation"
        style="color: rgb(248, 248, 242);"
      >
        (
      </span>
      <span
        class="token punctuation"
        style="color: rgb(248, 248, 242);"
      >
        )
      </span>
      <span
        class="token punctuation"
        style="color: rgb(248, 248, 242);"
      >
        ;
      </span>
    </div>
  </pre>
</div>
`;

exports[`BlockCode should render different language correctly (snapshot) 10`] = `
<div>
  <pre
    class="prism-code language-javascript code"
    style="color: rgb(118, 217, 230); background-color: rgb(42, 42, 42);"
  >
    <div
      class="token-line hover"
      style="color: rgb(118, 217, 230);"
    >
      <span
        class="number"
      >
        1
      </span>
      <span
        class="token keyword"
        style="color: rgb(249, 38, 114);"
      >
        const
      </span>
      <span
        class="token plain"
      >
         foo 
      </span>
      <span
        class="token operator"
        style="color: rgb(167, 122, 254);"
      >
        =
      </span>
      <span
        class="token plain"
      >
         
      </span>
      <span
        class="token function"
        style="color: rgb(166, 226, 45);"
      >
        bar
      </span>
      <span
        class="token punctuation"
        style="color: rgb(248, 248, 242);"
      >
        (
      </span>
      <span
        class="token punctuation"
        style="color: rgb(248, 248, 242);"
      >
        )
      </span>
      <span
        class="token punctuation"
        style="color: rgb(248, 248, 242);"
      >
        ;
      </span>
    </div>
  </pre>
</div>
`;

exports[`BlockCode should render different language correctly (snapshot) 11`] = `
<div>
  <pre
    class="prism-code language-ts code"
    style="color: rgb(118, 217, 230); background-color: rgb(42, 42, 42);"
  >
    <div
      class="token-line hover"
      style="color: rgb(118, 217, 230);"
    >
      <span
        class="number"
      >
        1
      </span>
      <span
        class="token keyword"
        style="color: rgb(249, 38, 114);"
      >
        const
      </span>
      <span
        class="token plain"
      >
         foo 
      </span>
      <span
        class="token operator"
        style="color: rgb(167, 122, 254);"
      >
        =
      </span>
      <span
        class="token plain"
      >
         
      </span>
      <span
        class="token function"
        style="color: rgb(166, 226, 45);"
      >
        bar
      </span>
      <span
        class="token punctuation"
        style="color: rgb(248, 248, 242);"
      >
        (
      </span>
      <span
        class="token punctuation"
        style="color: rgb(248, 248, 242);"
      >
        )
      </span>
      <span
        class="token punctuation"
        style="color: rgb(248, 248, 242);"
      >
        ;
      </span>
    </div>
  </pre>
</div>
`;

exports[`BlockCode should render different language correctly (snapshot) 12`] = `
<div>
  <pre
    class="prism-code language-typescript code"
    style="color: rgb(118, 217, 230); background-color: rgb(42, 42, 42);"
  >
    <div
      class="token-line hover"
      style="color: rgb(118, 217, 230);"
    >
      <span
        class="number"
      >
        1
      </span>
      <span
        class="token keyword"
        style="color: rgb(249, 38, 114);"
      >
        const
      </span>
      <span
        class="token plain"
      >
         foo 
      </span>
      <span
        class="token operator"
        style="color: rgb(167, 122, 254);"
      >
        =
      </span>
      <span
        class="token plain"
      >
         
      </span>
      <span
        class="token function"
        style="color: rgb(166, 226, 45);"
      >
        bar
      </span>
      <span
        class="token punctuation"
        style="color: rgb(248, 248, 242);"
      >
        (
      </span>
      <span
        class="token punctuation"
        style="color: rgb(248, 248, 242);"
      >
        )
      </span>
      <span
        class="token punctuation"
        style="color: rgb(248, 248, 242);"
      >
        ;
      </span>
    </div>
  </pre>
</div>
`;

exports[`BlockCode should render different language correctly (snapshot) 13`] = `
<div>
  <pre
    class="prism-code language-coffee code"
    style="color: rgb(118, 217, 230); background-color: rgb(42, 42, 42);"
  >
    <div
      class="token-line hover"
      style="color: rgb(118, 217, 230);"
    >
      <span
        class="number"
      >
        1
      </span>
      <span
        class="token plain"
      >
        const foo 
      </span>
      <span
        class="token operator"
        style="color: rgb(167, 122, 254);"
      >
        =
      </span>
      <span
        class="token plain"
      >
         
      </span>
      <span
        class="token function"
        style="color: rgb(166, 226, 45);"
      >
        bar
      </span>
      <span
        class="token punctuation"
        style="color: rgb(248, 248, 242);"
      >
        (
      </span>
      <span
        class="token punctuation"
        style="color: rgb(248, 248, 242);"
      >
        )
      </span>
      <span
        class="token punctuation"
        style="color: rgb(248, 248, 242);"
      >
        ;
      </span>
    </div>
  </pre>
</div>
`;

exports[`BlockCode should render different language correctly (snapshot) 14`] = `
<div>
  <pre
    class="prism-code language-coffeescript code"
    style="color: rgb(118, 217, 230); background-color: rgb(42, 42, 42);"
  >
    <div
      class="token-line hover"
      style="color: rgb(118, 217, 230);"
    >
      <span
        class="number"
      >
        1
      </span>
      <span
        class="token plain"
      >
        const foo 
      </span>
      <span
        class="token operator"
        style="color: rgb(167, 122, 254);"
      >
        =
      </span>
      <span
        class="token plain"
      >
         
      </span>
      <span
        class="token function"
        style="color: rgb(166, 226, 45);"
      >
        bar
      </span>
      <span
        class="token punctuation"
        style="color: rgb(248, 248, 242);"
      >
        (
      </span>
      <span
        class="token punctuation"
        style="color: rgb(248, 248, 242);"
      >
        )
      </span>
      <span
        class="token punctuation"
        style="color: rgb(248, 248, 242);"
      >
        ;
      </span>
    </div>
  </pre>
</div>
`;

exports[`BlockCode should render different language correctly (snapshot) 15`] = `
<div>
  <pre
    class="prism-code language-jsx code"
    style="color: rgb(118, 217, 230); background-color: rgb(42, 42, 42);"
  >
    <div
      class="token-line hover"
      style="color: rgb(118, 217, 230);"
    >
      <span
        class="number"
      >
        1
      </span>
      <span
        class="token keyword"
        style="color: rgb(249, 38, 114);"
      >
        const
      </span>
      <span
        class="token plain"
      >
         foo 
      </span>
      <span
        class="token operator"
        style="color: rgb(167, 122, 254);"
      >
        =
      </span>
      <span
        class="token plain"
      >
         
      </span>
      <span
        class="token function"
        style="color: rgb(166, 226, 45);"
      >
        bar
      </span>
      <span
        class="token punctuation"
        style="color: rgb(248, 248, 242);"
      >
        (
      </span>
      <span
        class="token punctuation"
        style="color: rgb(248, 248, 242);"
      >
        )
      </span>
      <span
        class="token punctuation"
        style="color: rgb(248, 248, 242);"
      >
        ;
      </span>
    </div>
  </pre>
</div>
`;

exports[`BlockCode should render different language correctly (snapshot) 16`] = `
<div>
  <pre
    class="prism-code language-tsx code"
    style="color: rgb(118, 217, 230); background-color: rgb(42, 42, 42);"
  >
    <div
      class="token-line hover"
      style="color: rgb(118, 217, 230);"
    >
      <span
        class="number"
      >
        1
      </span>
      <span
        class="token keyword"
        style="color: rgb(249, 38, 114);"
      >
        const
      </span>
      <span
        class="token plain"
      >
         foo 
      </span>
      <span
        class="token operator"
        style="color: rgb(167, 122, 254);"
      >
        =
      </span>
      <span
        class="token plain"
      >
         
      </span>
      <span
        class="token function"
        style="color: rgb(166, 226, 45);"
      >
        bar
      </span>
      <span
        class="token punctuation"
        style="color: rgb(248, 248, 242);"
      >
        (
      </span>
      <span
        class="token punctuation"
        style="color: rgb(248, 248, 242);"
      >
        )
      </span>
      <span
        class="token punctuation"
        style="color: rgb(248, 248, 242);"
      >
        ;
      </span>
    </div>
  </pre>
</div>
`;

exports[`BlockCode should render different language correctly (snapshot) 17`] = `
<div>
  <pre
    class="prism-code language-objc code"
    style="color: rgb(118, 217, 230); background-color: rgb(42, 42, 42);"
  >
    <div
      class="token-line hover"
      style="color: rgb(118, 217, 230);"
    >
      <span
        class="number"
      >
        1
      </span>
      <span
        class="token keyword"
        style="color: rgb(249, 38, 114);"
      >
        const
      </span>
      <span
        class="token plain"
      >
         foo 
      </span>
      <span
        class="token operator"
        style="color: rgb(167, 122, 254);"
      >
        =
      </span>
      <span
        class="token plain"
      >
         
      </span>
      <span
        class="token function"
        style="color: rgb(166, 226, 45);"
      >
        bar
      </span>
      <span
        class="token punctuation"
        style="color: rgb(248, 248, 242);"
      >
        (
      </span>
      <span
        class="token punctuation"
        style="color: rgb(248, 248, 242);"
      >
        )
      </span>
      <span
        class="token punctuation"
        style="color: rgb(248, 248, 242);"
      >
        ;
      </span>
    </div>
  </pre>
</div>
`;

exports[`BlockCode should render different language correctly (snapshot) 18`] = `
<div>
  <pre
    class="prism-code language-objectivec code"
    style="color: rgb(118, 217, 230); background-color: rgb(42, 42, 42);"
  >
    <div
      class="token-line hover"
      style="color: rgb(118, 217, 230);"
    >
      <span
        class="number"
      >
        1
      </span>
      <span
        class="token keyword"
        style="color: rgb(249, 38, 114);"
      >
        const
      </span>
      <span
        class="token plain"
      >
         foo 
      </span>
      <span
        class="token operator"
        style="color: rgb(167, 122, 254);"
      >
        =
      </span>
      <span
        class="token plain"
      >
         
      </span>
      <span
        class="token function"
        style="color: rgb(166, 226, 45);"
      >
        bar
      </span>
      <span
        class="token punctuation"
        style="color: rgb(248, 248, 242);"
      >
        (
      </span>
      <span
        class="token punctuation"
        style="color: rgb(248, 248, 242);"
      >
        )
      </span>
      <span
        class="token punctuation"
        style="color: rgb(248, 248, 242);"
      >
        ;
      </span>
    </div>
  </pre>
</div>
`;

exports[`BlockCode should render different language correctly (snapshot) 19`] = `
<div>
  <pre
    class="prism-code language-vue code"
    style="color: rgb(118, 217, 230); background-color: rgb(42, 42, 42);"
  >
    <div
      class="token-line hover"
      style="color: rgb(118, 217, 230);"
    >
      <span
        class="number"
      >
        1
      </span>
      <span
        class="token plain"
      >
        const foo = bar();
      </span>
    </div>
  </pre>
</div>
`;

exports[`BlockCode should render different language correctly (snapshot) 20`] = `
<div>
  <pre
    class="prism-code language-go code"
    style="color: rgb(118, 217, 230); background-color: rgb(42, 42, 42);"
  >
    <div
      class="token-line hover"
      style="color: rgb(118, 217, 230);"
    >
      <span
        class="number"
      >
        1
      </span>
      <span
        class="token keyword"
        style="color: rgb(249, 38, 114);"
      >
        const
      </span>
      <span
        class="token plain"
      >
         foo 
      </span>
      <span
        class="token operator"
        style="color: rgb(167, 122, 254);"
      >
        =
      </span>
      <span
        class="token plain"
      >
         
      </span>
      <span
        class="token function"
        style="color: rgb(166, 226, 45);"
      >
        bar
      </span>
      <span
        class="token punctuation"
        style="color: rgb(248, 248, 242);"
      >
        (
      </span>
      <span
        class="token punctuation"
        style="color: rgb(248, 248, 242);"
      >
        )
      </span>
      <span
        class="token punctuation"
        style="color: rgb(248, 248, 242);"
      >
        ;
      </span>
    </div>
  </pre>
</div>
`;

exports[`BlockCode should render different language correctly (snapshot) 21`] = `
<div>
  <pre
    class="prism-code language-java code"
    style="color: rgb(118, 217, 230); background-color: rgb(42, 42, 42);"
  >
    <div
      class="token-line hover"
      style="color: rgb(118, 217, 230);"
    >
      <span
        class="number"
      >
        1
      </span>
      <span
        class="token plain"
      >
        const foo = bar();
      </span>
    </div>
  </pre>
</div>
`;

exports[`BlockCode should render different language correctly (snapshot) 22`] = `
<div>
  <pre
    class="prism-code language-rust code"
    style="color: rgb(118, 217, 230); background-color: rgb(42, 42, 42);"
  >
    <div
      class="token-line hover"
      style="color: rgb(118, 217, 230);"
    >
      <span
        class="number"
      >
        1
      </span>
      <span
        class="token plain"
      >
        const foo = bar();
      </span>
    </div>
  </pre>
</div>
`;
